src/Messages/BookingCodesMessage.php

```diff
@@ -0,0 +1,195 @@
+
+        $bookingTable=$this->getCodes($bookingCodes);
+        if(empty($bookingTable['table'])) {
+            do_action( 'commonsbooking_mail_sent', $this->getAction(),
+                        new \WP_Error(
+                            "e-mail",
+                            commonsbooking_sanitizeHTML( __( "Could not find Booking codes for this timeframe/period", "commonsbooking" ) ),
+                            [ 'title' => commonsbooking_sanitizeHTML( __( "Error sending booking codes", "commonsbooking" ) ) ]
+                        )
+            );
+            
+            return false;
+        }
+
+        //add Table content to post_meta of item for template rendering
+        update_post_meta( $item->ID, '_codeTable', $bookingTable['table'] );
+        
+        $dispFrom= wp_date("M-Y",$bookingTable['startDate']);
+        $dispTo= wp_date("M-Y",$bookingTable['endDate']);
+        if($dispFrom == $dispTo)
+            update_post_meta( $item->ID, '_dateRange', $dispFrom );
+        else
+            update_post_meta( $item->ID, '_dateRange', $dispFrom . " - " . $dispTo );
+
+		// get templates from Admin Options
+		$template_body    = Settings::getOption( 'commonsbooking_options_templates',
+			'emailtemplates_mail-booking-' . $this->action . '-body' );
+        $template_subject = Settings::getOption( 'commonsbooking_options_templates',
+			'emailtemplates_mail-booking-' . $this->action . '-subject' );
+        $template_bcc = sanitize_email(Settings::getOption( 'commonsbooking_options_templates',
+			'email_mail-booking-' . $this->action . '-bcc' ));
+
+		// Setup email: From
+		$fromHeaders = sprintf(
+			"From: %s <%s>",
+			Settings::getOption( 'commonsbooking_options_templates', 'emailheaders_from-name' ),
+			sanitize_email( Settings::getOption( 'commonsbooking_options_templates', 'emailheaders_from-email' ) )
+		);
+
+		$this->prepareMail(
+			$itemAdmins[0],
+			$template_body,
+			$template_subject,
+			$fromHeaders,
+			$template_bcc,
+			[
+                'item' => $item,
+                'location' => $timeframe->getLocation(),
+            ]
+		);
+        
+        //cleanup post_meta thats only used for template rendering
+        delete_post_meta( $item->ID, '_codeTable');
+        delete_post_meta( $item->ID, '_dateRange');
+
+        //in the case of multiple receivers add real to: because base class does not allow multiple users
+        if(count($itemAdmins) > 1)
+        {
+            $this->to=array();
+            foreach($itemAdmins as $admin)
+                $this->to[]=sprintf( '%s <%s>', $admin->user_nicename, $admin->user_email );
+        }
+
+
+		$this->SendNotificationMail();
+
+        return true;
+    }
+
+ 	/**
+	 * renders Booking Codes Table
+	 *
+     * @param  array $bookingCodes    array of Booking Codes for this timeframe.
+     *
+     * @return array                  HTML of table and dates of first and last Code.
+ 	 */
+      protected function getCodes($bookingCodes): array {
```

Comment:
Jetzt gibt es zwei Tabellen die generiert werden, eine im Backend und eine für die E-Mail. Könnte man nicht in View die renderTable Funktion verwenden? Vielleicht indem sie so umgeschrieben wird, dass sie kein echo macht sondern einen String zurückgibt?